src/fsharp/Interactive.DependencyManager/DependencyManager.fs

```diff
@@ -0,0 +1,246 @@
+            if isNull property then
+                None
+            elif not (property.GetGetMethod().IsStatic)
+                 && property.GetIndexParameters() = Array.empty
+            then
+                Some property
+            else
+                None
+        with | _ -> None
+
+    let getInstanceMethod<'treturn> (theType: Type) (parameterTypes: Type array) methodName =
+        try
+            let theMethod = theType.GetMethod(methodName, parameterTypes)
+            if isNull theMethod then
+                None
+            else
+                Some theMethod
+        with | _ -> None
+
+    let stripTieWrapper (e:Exception) =
+        match e with
+        | :? TargetInvocationException as e->
+            e.InnerException
+        | _ -> e
+
+open ReflectionHelper
+
+(* Shape of Dependency Manager contract, resolved using reflection *)
+[<AllowNullLiteralAttribute>]
+type IDependencyManagerProvider =
+    abstract Name: string
+    abstract Key: string
+    abstract ResolveDependencies: scriptDir: string * mainScriptName: string * scriptName: string * scriptExt: string * packageManagerTextLines: string seq * tfm: string -> bool * string seq * string seq * string seq
+
+[<RequireQualifiedAccess>]
+type ErrorReportType =
+| Warning
+| Error
+
+type ReflectionDependencyManagerProvider(theType: Type, nameProperty: PropertyInfo, keyProperty: PropertyInfo, resolveDeps: MethodInfo option, resolveDepsEx: MethodInfo option,outputDir: string option) =
+    let instance = Activator.CreateInstance(theType, [|outputDir :> obj|])
+    let nameProperty     = nameProperty.GetValue >> string
+    let keyProperty      = keyProperty.GetValue >> string
+
+    static member InstanceMaker (theType: System.Type, outputDir: string option) =
+        match getAttributeNamed theType dependencyManagerAttributeName,
+              getInstanceProperty<string> theType namePropertyName,
+              getInstanceProperty<string> theType keyPropertyName
+              with
+        | None, _, _
+        | _, None, _
+        | _, _, None -> None
+        | Some _, Some nameProperty, Some keyProperty ->
+            let resolveMethod = getInstanceMethod<bool * string list * string list> theType [| typeof<string>; typeof<string>; typeof<string>; typeof<string seq>; typeof<string> |] resolveDependenciesMethodName
+            let resolveMethodEx = getInstanceMethod<bool * string list * string list> theType [| typeof<string>; typeof<string seq>; typeof<string> |] resolveDependenciesMethodName
+            Some (fun () -> new ReflectionDependencyManagerProvider(theType, nameProperty, keyProperty, resolveMethod, resolveMethodEx, outputDir) :> IDependencyManagerProvider)
+
+    interface IDependencyManagerProvider with
+
+        /// Name of dependency Manager
+        member __.Name     = instance |> nameProperty
+
+        /// Key of dependency Manager: used for #r "key: ... "   E.g nuget
+        member __.Key      = instance |> keyProperty
```

Comment:
These don't appear to be aligning with anything. Why the extra spaces?

Reply:
I am guessing, I was bored or maybe fell asleep at the keyboard :-)